### PR TITLE
SslBump: Support parsing GREASEd (and future) TLS handshakes

### PR DESCRIPTION
**Pull request by chtsanti**
A peeking or staring Squid aborted TLS connections containing a GREASE
version in the supported_versions handshake extension (e.g., 0x3a3a).
Here is a sample cache.log error (debug_options ALL,1 83,2):

```
parseTlsHandshake: ... check failed: vMajor == 3
    exception location: Handshake.cc(119) ParseProtocolVersion
```

The same problem would apply to some other "unsupported" (and currently
non-existent) TLS versions (e.g., 0x0400). The growing popularity of
GREASE values exposed the bug (just like RFC 8710 was designed to do).

Squid now ignores all unsupported-by-Squid TLS versions in handshake
extensions. Squid still requires a supported TLS version for
framing-related handshake fields -- no changes there.

It is difficult to define "supported" in this context precisely because
a peeking Squid only observes the TLS handshake. Our handshake parser
may report the following versions: SSL v2.0, SSL v3.0, and TLS v1.x
(with x >= 0). This logic allows us to safely parse the handshake (no
framing errors) and continue to make version-based decisions like
disabling OpenSSL TLS v1.3 support when the agents are negotiating TLS
v1.3+ (master cd29a42). Also, access logs benefit from this "support"
going beyond TLS v1.2 even though Squid cannot bump most TLS v1.3+
handshakes.

Squid was and still is compliant with the "MUST NOT treat GREASE values
differently from any unknown value" requirement of RFC 8710.

Also added source code comments to mark other places where unsupported
(for some definition of "support") TLS values, including GREASE values
may appear.

Folks studying debugging logs (where GREASE values may appear) will
probably either recognize their easy-to-remember 0x?A?A pattern or
not really know/care about RFC 8710 and its special values.

This is a Measurement Factory project.

### CODE REVIEW

File: src/security/Handshake.cc [hunk 1]
```diff
@@ -554,12 +583,19 @@ Security::HandshakeParser::parseSupportedVersionsExtension(const SBuf &extension
         Parser::BinaryTokenizer tkList(extensionData);
         Parser::BinaryTokenizer tkVersions(tkList.pstring8("SupportedVersions"));
         while (!tkVersions.atEnd()) {
-            const auto version = ParseProtocolVersion(tkVersions, "supported_version");
+            const auto version = ParseOptionalProtocolVersion(tkVersions, "supported_version");
+            // Ignore values unsupported by Squid. Today, those are likely to be
+            // RFC 8701 GREASE values. Eventually, we may start seeing TLS v2+.
```
[the remark this answers is not in the document]
Comment (rousskov) [reply src/security/Handshake.cc:588]: The documented act of "ignoring" happens in _this_ function. The parser decides what versions are unsupported (and morphs them all together), but it does _not_ make the decision whether to ignore unsupported versions. I can imagine a different ParseOptionalProtocolVersion() caller that would do something other than simply ignoring unsupported versions.

If you insist, we will move this "Ignore..." documentation to the parser function (that ignores nothing), but I think it is currently in the best place possible.

File: src/security/Handshake.cc [hunk 2]
```diff
@@ -104,25 +105,49 @@ class Extension
...
+        if (vMinor == 0)
+            return AnyP::ProtocolVersion(AnyP::PROTO_SSL, 3, 0);
+        return AnyP::ProtocolVersion(AnyP::PROTO_TLS, 1, (vMinor - 1));
+    }
+
+    const uint16_t vRaw = (vMajor << 8) | vMinor;
+    debugs(83, 7, "unsupported: " << asHex(vRaw));
+    if (beStrict)
+        throw TextException(ToSBuf("unsupported TLS version: ", asHex(vRaw)), Here());
+    return AnyP::ProtocolVersion();
```
Comment (yadij) [inline src/security/Handshake.cc:130]:
```suggestion
    debugs(83, 7, "unsupported: " << asHex(vRaw));
    if (beStrict)
        throw TextException(ToSBuf("unsupported TLS version: ", asHex(vRaw)), Here());
    // else, ignore unknown and RFC 8701 values
    return AnyP::ProtocolVersion();
```

File: src/security/Handshake.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/security/Handshake.cc:588]: I've added a comment above where the parser ignores these GREASE values in non-strict code path.

**Comment by yadij** [comment]
0ddbd4e  is a good example of why having the GREASE documentation outside ParseOptionalProtocolVersion() is bad. It forgets to update comments on the other callers doing version ignore.

**Comment by rousskov** [comment]
> 0ddbd4e is a good example of why having the GREASE documentation outside ParseOptionalProtocolVersion() is bad. It forgets to update comments on the other callers doing version ignore.

It is actually the opposite: The other caller (i.e. the one handling from-server versions) should not see GREASE values. The difference (in documentation of what we expect each caller to see) is _meaningful_, not accidental.

However, if you insist, I can remove references to GREASE values from parser and/or its caller. Removing caller documenation would make the code a little more difficult to understand/maintain, IMO, but it is not such a big deal. Just let me know if you want me to delete those references or delete them yourself.

**Comment by yadij** [comment]
I think it better not to mention GREASE at all. The point of the GREASE initiative is to break software which is restricting TLS extensibility points - forcing bug fixes to happen. In this case TLS/SSL version numbers are extensible - implementations SHOULD be ignoring unsupported versions in the handshake extension field even if the GREASE RFC did not exist.

**Comment by rousskov** [comment]
> I think it better not to mention GREASE at all.

Done at 1f9417f.

FWIW, I think hiding the truth about special values from developers is a mistake, especially when that information can be clearly expressed in just a few words. The facts you provided do not substantiate your "do not mention" opinion AFAICT. However, wasting even more time arguing about these comments is counterproductive IMO.

**Pull request closed without merging**